### ISSUE TITLE
java.lang.ClassCastException: java.lang.invoke.BoundMethodHandle$Species_LL cannot be cast to java.lang.invoke.SimpleMethodHandle

### ISSUE THREAD

**Issue by etspaceman** [oracle/graal#3341]
**Describe the issue**
Something seems to be missing from the MethodHandle support, getting:

```
Caused by: java.lang.ClassCastException: java.lang.invoke.BoundMethodHandle$Species_LL cannot be cast to java.lang.invoke.SimpleMethodHandle
	at com.oracle.svm.methodhandles.MethodHandleIntrinsic.execute(MethodHandleIntrinsic.java:254)
	at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:80)
	at java.lang.invoke.LambdaForm$NamedFunction.invokeWithArguments(LambdaForm.java:74)
	at java.lang.invoke.LambdaForm.interpretName(LambdaForm.java:981)
	at java.lang.invoke.LambdaForm.interpretWithArguments(LambdaForm.java:958)
	at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:171)
	at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:0)
	at java.lang.invoke.LambdaForm$MH/1417001346.invokeExact_MT(LambdaForm$MH)
	at org.apache.lucene.store.MMapDirectory.lambda$null$0(MMapDirectory.java:404)
	at java.security.AccessController.doPrivileged(AccessController.java:84)
	at org.apache.lucene.store.MMapDirectory.lambda$newBufferCleaner$1(MMapDirectory.java:402)
	... 47 more
```
Creating based on : 

https://github.com/oracle/graal/issues/2761#issuecomment-785006081

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. Clone https://github.com/etspaceman/kinesis-mock/tree/4acc3ee60a6dda9ea85629c2551a25faa9985ac6
2. Run `sbt dockerComposeUp`

**Describe GraalVM and your environment:**
 - Using image ghcr.io/graalvm/graalvm-ce:java11-21.0.0.2
 
[message.txt](https://github.com/oracle/graal/files/6298932/message.txt)

**Comment by mcraj017** [comment]
@etspaceman thanks for the issue
we will check it out and get back to you

**Comment by dainiusjocas** [comment]
I have the same issue.

**Comment by mcraj017** [comment]
@etspaceman when I tried your reproducer, I didn't get any error:
```
docker run ghcr.io/etspaceman/kinesis-mock:0.0.5-SNAPSHOT
[ioapp-compute-1] INFO  k.m.KinesisMockService contextId=c39ba88d-a8d1-11eb-87b3-176e2a535263, cacheConfig={"awsAccountId":"000000000000","awsRegion":"us-east-1","createStreamDuration":"500 milliseconds","deleteStreamDuration":"500 milliseconds","deregisterStreamConsumerDuration":"500 milliseconds","mergeShardsDuration":"500 milliseconds","registerStreamConsumerDuration":"500 milliseconds","shardLimit":50,"splitShardDuration":"500 milliseconds","startStreamEncryptionDuration":"500 milliseconds","stopStreamEncryptionDuration":"500 milliseconds","updateShardCountDuration":"500 milliseconds"} - Logging Cache Config 
[ioapp-compute-2] INFO  k.m.KinesisMockService  - Starting Kinesis Http2 Mock Service on port 4567 
[ioapp-compute-2] INFO  k.m.KinesisMockService  - Starting Kinesis Http1 Plain Mock Service on port 4568 
```

**Comment by mcraj017** [comment]
@etspaceman please share any other information which can help me to reproduce this issue

**Comment by kaspernielsen** [comment]
I'm seeing the same issue.

```
public class HelloWorld {

    static final Helper helper;

    static {
        helper = Helper.scan(HelloWorld.class);

    }

    static void ffooo() {
        System.out.println("Works fine");
    }

    public static void main(String[] args) throws Throwable {
        helper.run();
    }

    static record Helper(MethodHandle mh) {

        Helper(MethodHandle mh) {
            this.mh = MethodHandles.dropArguments(mh, 0, String.class);
        }

        public void run() throws Throwable {
            mh.invokeExact("dada");
        }

        static Helper scan(Class<?> cl) {
            for (Method m : cl.getDeclaredMethods()) {
                if (m.getParameterCount() == 0) {
                    Lookup l = MethodHandles.lookup();
                    try {
                        return new Helper(l.unreflect(m));
                    } catch (IllegalAccessException e) {
                        throw new Error(e);
                    }
                }
            }
            throw new UnsupportedOperationException();
        }

    }
}
```
Compiled with native-image HelloWorld --initialize-at-build-time=HelloWorld foobar

**Comment by mcraj017** [comment]
@kaspernielsen thanks for the reproducer, I am able to reproduce the issue
I will raise to our Dev team for further steps

**Comment by etspaceman** [comment]
Thanks for the reproducer @kaspernielsen . @mcraj017 - I switched out the library that was causing my issue in my latest versions. The commit that I had linked had this issue. But it seems we have a much simpler reproducer here which is good.

**Comment by christianwimmer** [comment]
Is this still reproducible with GraalVM 21.1? We definitely fixed a bug recently that caused a similar exception.

**Comment by kaspernielsen** [comment]
The snippet I posted failed with 21.1

**Comment by moorsu** [comment]
I am also encountering the same issue with 21.1

```
Exception in thread "main" java.lang.ClassCastException: java.lang.invoke.BoundMethodHandle$Species_LLLLLLLLLLLLL cannot be cast to java.lang.invoke.SimpleMethodHandle
	at com.oracle.svm.methodhandles.MethodHandleIntrinsicImpl.execute(MethodHandleIntrinsicImpl.java:258)
	at com.oracle.svm.methodhandles.Util_java_lang_invoke_MethodHandle.invokeInternal(Target_java_lang_invoke_MethodHandle.java:144)
	at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:79)
	at java.lang.invoke.LambdaForm$NamedFunction.invokeWithArguments(LambdaForm.java:76)
	at java.lang.invoke.LambdaForm.interpretName(LambdaForm.java:981)
	at java.lang.invoke.LambdaForm.interpretWithArguments(LambdaForm.java:958)
	at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:85)
	at java.lang.invoke.MethodHandle.invokeBasic(MethodHandle.java:0)
	at java.lang.invoke.LambdaForm$MH/15040032.invoke(LambdaForm$MH)
```

**Comment by moorsu** [comment]
Hi  loicottet, Did you get a chance to look into it?.  Thanks in advance!

**Comment by loicottet** [comment]
The fix is ready and should be merged to master soon.

**Comment by loicottet** [comment]
Fixed by 540eff5e0b6edf7cffafba9fad5c874446d5299e
